EOMBase class migration of property attributes

jbm950 opened the pull request:
I figured as the equations of motion base class is being developed, it'd be good to have this location to discuss the desired api and such. Currently I only have some pseudo test code and nothing else.

jbm950 committed:
Added a file for EOMBase class and test_eombase

The eombase.py is currently empty and the test file has some pseudo test
code.

moorepants commented on sympy/physics/mechanics/tests/test_eombase.py:
```diff
@@ -0,0 +1,30 @@
+import sympy.physics.mechanics.eombase as eombase
+
+# Determine what inputs should be used
+EOM = eombase.EOMBase(frame=None)
```
We currently have `LagrangesMethod` and `KanesMethod` classes. These classes take descriptions of multibody dynamic systems in the form of generalized coordinates/speeds, kinematically defined bodies, and constraints and produce symbolic equations of motion, i.e. it automates the EoM formation method.

We'd like to support future methods, e.g. `HamiltonMethod`, `NewtonEulerMethod`, `FeatherstoneMethod`, `DenavitHartenbergMethod` etc. We'd also like to support any general "manual" method. This could mean that the user finds the EoMs written already for them, or that they write them out by hand and then type them in SymPy code, or that they use SymPy to do symbolic computations for their custom method. The results of any of these should be everything that fully describes the symbolic form of the equations of motion.

One thought: Should we have a class that describes "equations of motion", and if so should the `*Method` classes produce an instance of this?

One thing that is nice about the current `*Method` classes is that they contain references to the bodies and loads. Should we care about that?

sksingh1202 replied:
Hello,
I have gone through the gsoc 2020 ideas list and was interested in creating a new class for NewtonEuler's method and Hamilton's method. I am new to Sympy's codebase. I have already gone through Sympy's tutorial and the Introduction to contributing page. Could you suggest me on how could I start to contribute.
Thank you.

moorepants replied:
Start here: https://github.com/sympy/sympy/wiki/introduction-to-contributing

sksingh1202 replied:
Thank you very much.

moorepants commented on sympy/physics/mechanics/tests/test_eombase.py:
```diff
@@ -0,0 +1,30 @@
+
+Thing = "Place holder until a correct output is chosen"
+
+assert EOM.auxiliary_eqs == Thing
```
This is only relevant to Kane's method, I think. Probably shouldn't be in the base class.

moorepants commented on sympy/physics/mechanics/tests/test_eombase.py:
```diff
@@ -0,0 +1,30 @@
+assert EOM.forcing == Thing
+assert EOM.forcing_full == Thing
+assert EOM.q == Thing
+assert EOM.u == Thing
```
Here is what I see as the essentials to describe the EoMs of a multibody system:

Option 1

1. The system states, `x`.
2. The first order ordinary differential equation, i.e the right hand side of `x' = f`.

For a multibody system it may be worth knowing which entries of `f` are associated with the generalized speeds and which are associated with the generalized coordinates.

Option 2
1. The system states, `x`.
2. The implicit first order ordinary differential equation in the form `M x' = F`.

Lots more...will add later.

moorepants commented on sympy/physics/mechanics/tests/test_eombase.py:
```diff
@@ -0,0 +1,30 @@
+assert EOM.forcelist == Thing
+assert EOM.forcing == Thing
+assert EOM.forcing_full == Thing
+assert EOM.q == Thing
```
I suggest more explicit names: coordinates, speeds instead of q, u, etc.

moorepants commented:
I recommend reading this PR: https://github.com/pydy/pydy/pull/345

moorepants commented on sympy/physics/mechanics/tests/test_eombase.py:
```diff
@@ -0,0 +1,30 @@
+Thing = "Place holder until a correct output is chosen"
+
+assert EOM.auxiliary_eqs == Thing
+assert EOM.bodylist == Thing
```
I think `bodies` is a better name. It probably should default to a tuple, not a list and it should work with any iterable.

moorepants commented on sympy/physics/mechanics/tests/test_eombase.py:
```diff
@@ -0,0 +1,30 @@
+assert EOM.bodylist == Thing
+assert EOM.mass_matrix == Thing
+assert EOM.mass_matrix_full == Thing
+assert EOM.forcelist == Thing
```
I think `loads` is a better name because it encompasses both "forces" and "torques".

jbm950 commented:
`"Should we have a class that describes "equations of motion", and if so should the *Method classes produce an instance of this?"`

Hadn't thought about this. I guess a simple `EOM` object could be produced by the *Method classes. If this is the direction we decide to go would a base class be necessary? The *Method classes could exclusively contain the code to produce the equations of motion and the `EOM` object could be where the results are stored, accessed and processed (code such as the linearize method and rhs methods could be removed from the *Method classes). Obviously this would be much less backwards compatible than creation of a parent class for the *Method classes and would thus require many deprecation warnings. 

As far as promoting addition of new methods I believe the main benefit that would be gained is that there would be a standard expected output that is guaranteed to work with the rest of the code base. Also, the *Methods can have rather different methods of eom generation and may not necessarily benefit from having stock code aimed at the generation of the eom themselves (for instance Featherstone's method uses 6d spatial vectors and wouldn't need the 3D coordinates that the other classes would). Thus the main benefit from having a base class would be the storage and processing of the eom.

`"This is only relevant to Kane's method, I think. Probably shouldn't be in the base class."` (in reference to the auxiliary equations)

pydy.system.System pulls out auxiliary `u`'s in its `_kane_inlist_syms` method. Would this be something that Kane's Method should allow access to, however, rather than be needed in the base class? For instance if using a `EOM` class like discussed above, the auxiliary information would be processed before storing into `EOM` rather than storing it separately in `EOM`?

`"I suggest more explicit names: coordinates, speeds instead of q, u, bodies, loads"`

My goal was to promote backwards compatibility as much as possible (for instance moving the q and u properties entirely from `KanesMethod` and `LagrangesMethod` to the base class. A change in names would be beneficial for long term maintenance but would require deprecation warnings if I went ahead with migrating that functionality to the new class.

jbm950 committed:
Mock up of EOM class

Created a generic mock up of an equations of motion class. This class
would not be inherited from by the *Methods classes but would be a
returned object when the equations of motion are formed.

jbm950 committed:
Changed some attribute names in the eom test code

Changed the names of some attributes that the eombase test code expects
to be more explicit as Jason suggested.

jbm950 commented:
I'm looking at inputting the dynamical differential equations and the kinematical differential equations into the EOM class upon initialization. I have found a difference in how `KanesMethod` and `LagrangesMethod` stores this information. In `LagrangesMethod` the dynamical equations are stored such that they solve `Kd*udot = Fd` but in `KanesMethod` they are stored such that `Kd*udot + Fd = 0`. This means that Fd(Kanes) = - Fd(Lagranges). When returning the forcing vector `KanesMethod` multiplies Fd by -1 thus moving it to the other side of the equal sign. Which would be a better method of input into the EOM class? 

I'm thinking about having it input such that it satisfies the `Kd*udot = Fd` form. I was wondering though if this means I should input the kinematical equation in this form also with most terms moved to the rhs of the equal sign. (`Kkqdot*qdot = Kku * u + Fk` instead of `Kkqdot*qdot = -(Kku * u + Fk)`)

jbm950 committed:
Added real test code for EOM

I have changed the pseudo code to real test code. Currently the code is
still messy and has extraneous bits (such as the formation of a
LagrangesMethod instance).

jbm950 committed:
Created a prototype of the EOM class

Filled in the methods such that the test code now completely passes.

jbm950 commented:
`"I guess a simple EOM object could be produced by the *Method classes. If this is the direction we decide to go would a base class be necessary? The *Method classes could exclusively contain the code to produce the equations of motion and the EOM object could be where the results are stored, accessed and processed (code such as the linearize method and rhs methods could be removed from the *Method classes). Obviously this would be much less backwards compatible than creation of a parent class for the *Method classes and would thus require many deprecation warnings."`

I've put more time into thinking about the benefits of this vs a base class

Advantages of returnable object

- Fewer additional methods attached to the object that other code will be using
- *Method classes can focus exclusively on generating the EOM without having to worry about the desired interface for using the EOM

Advantages of base class

- Some *Methods have additional parameters/methods that are specific to that method and would be better as an extension of a class rather than trying to add additional specific methods to a returnable class
    - Examples
        - `KanesMethod` auxiliary equations/speeds
        - `LagrangesMethod` solve multipliers method
    - Easer to extend

In order to help decide between the two solutions a greater definition of the goal of the project would be helpful. In general the goal of the project is to make it "easier" to add other equations of motion generators to be added to the SymPy repository. Different equations of motion generators are sufficiently varied in their operation that I do not think some "base code" will be broadly applicable for generation. Therefore I believe "easier" in this sense will relate more to using the EOM that are generated by whichever method. Thus the project I believe should be focused more on a unified way for generated equations of motion to be used (the interface after generation).  To this end both of the above solutions are feasible as this is the aspect that both of them touch. I believe the second solution (base/parent class) will be the better solution, however, as it is easier to extend between the two which allow per *Method specifics to be specified. As such I am going to change the code presented here in to fit the base class solution.

Summary

Using a parent class approach because it can still achieve a universal interface for different equations of motion generation while at the same time is able to be extended to provide method specific information.

jbm950 commented:
It was suggested to change some variables to more explicit names. When implementing the code as a parent class should the previous names be deprecated to match the name change?

ie.
`~~.q  ->  ~~.coordinates`
`~~.u  ->  ~~.speeds`
`~~.forcelist  ->  ~~.loads`
`~~.bodylist ->  ~~.bodies`

If so would the preference be of the semi-private versions of those variables be (`~~._q`)? Should these be changed as well?

jbm950 commented:
@sympy/mechanics

jbm950 committed:
Cleaned up test code

Removed some extraneous code from test_eombase.py

jbm950 committed:
Migrated property attributes to a base class

I migrated the property attributes from the equations of motion classes
to a base class. This will ensure that the equations of motion should
have the same basic interface for accessing the EOM regardless of which
method is used and will begin to reduce code redundancy.

I also added a method to the base class to check if the equations of
motion have been calculated yet. This check appeared in both methods as
copy pasted lines of code and so this has now been centralized for
convienence.

jbm950 commented:
To narrow the scope of this specific pull request I have decided to limit it to specifically migrating the property attributes to the base class. These property attributes I see as the main method of interfacing with the equations of motion generators and so having them present in a base class helps ensure a uniform interface between the *Methods. Other changes such as variable name changing I felt would be better suited as a separate pull request.

moorepants commented:
Some responses:

>  If this is the direction we decide to go would a base class be necessary?

Maybe not.

> Obviously this would be much less backwards compatible than creation of a parent class for the *Method classes and would thus require many deprecation warnings.

I'd like to avoid deprecation, but if the change in design is worth it, then it is ok. Maybe sketch out the new design so we could decide.

> for instance Featherstone's method uses 6d spatial vectors and wouldn't need the 3D coordinates that the other classes would

I think the user should be allowed to form the kinematics the same was as usual, but the 3D vectors should be converted internally in Featherstone's method into 6D for the calculations. This is essentially how Simbody handles it. The 6D vector object can still be available for explicit use and maybe be an option to pass into the Featherstone method.

> Would this be something that Kane's Method should allow access to, however, rather than be needed in the base class?

System needs to be able to handle "output equations", i.e. algebraic equations that are functions of the states and parameters. The auxiliary speeds are used to compute internal forces, one example of additional outputs. Other things like kinetic energy and potential energy are common things that are computed. Ignore the stuff that has to do with auxiliary in KanesMethod, as we need to design for general output equations and this would fall into that realm. All methods have this issue.

>  A change in names would be beneficial for long term maintenance but would require deprecation warnings if I went ahead with migrating that functionality to the new class.

Just support the old and the new. In Kane's method `q` will be an alias for `coordinates`. I've started some work on cleaning up Kane's Method along with my PR with the docs fix.

>  In LagrangesMethod the dynamical equations are stored such that they solve Kd*udot = Fd but in KanesMethod they are stored such that Kd*udot + Fd = 0

KanesMethod is based on "mass_matrix * u' = forcing". I think Lagrange is the same.

>  I was wondering though if this means I should input the kinematical equation in this form also with most terms moved to the rhs of the equal sign.

We should choose some common ways of storing these. KanesMethod stores the kinematical equations in a dictionary that maps the q' to the expression. But internally it also has the matrix form as non-public attributes.

> Using a parent class approach because it can still achieve a universal interface for different equations of motion generation while at the same time is able to be extended to provide method specific information.

I'd like to see a sketched out examples of what the API would look like for both cases. I think that a DifferentialEquation class along with ODE and DAE subclasses could likely make things really clean, especially for passing to System in pydy. It would be much more general.

I'm not sure our linearizer will work with very general ODE and DAEs, it may need some special information about what are independent and dependent coordinates. But it can be generalized.

A method on th e`*Method` classes that produces the general ODE and DAE forms would be nice, we could keep things backwards compatible in System so that it accepts a `*Method` or `ODE` or `DAE`.

>  When implementing the code as a parent class should the previous names be deprecated to match the name change?

Just support both as aliases.

moorepants commented:
One suggestion:

Don't spend any time writing the implementation, just write tests. Tests will very explicitly describe the desired API and functionality. We can then review these tests to come to an agreement of the new API design. Once the API is agreed upon, then implement. Otherwise you'll spend time implementing stuff that may not make it through the API design process.

moorepants commented:
With regards to the last comment, Sampad did a nice job laying out a proposed API here: https://github.com/sympy/sympy/wiki/GSoC-2016-Application-Sampad-Kumar-Saha:-Singularity-Functions for his project. We've discussed it and modified it over the past month and now he's just implementing (after writing the tests). It's going pretty smoothly. I think it would be nice to have something similar here. We could compare competing interfaces. It would be nice to see a theorectical python session that goes through and example problem from EoM formation to System calls.

moorepants commented:
This could also warrant a video chat. Have you an Angadh discussed this in a meeting yet?

moorepants commented:
Reference for simple DAE problem: http://nbviewer.jupyter.org/github/bmcage/odes/blob/master/docs/ipython/Planar%20Pendulum%20as%20DAE.ipynb

We need system to work for this. We have this same problem in our lagrange tests. Ref: https://github.com/pydy/pydy/issues/127

jbm950 commented:
Hi Jason,

Angadh and I have not yet discussed this in a meeting. I can meet tomorrow morning if you two are available. @angadhn 

Perhaps to go in-depth with an api discussion, a wiki page like the link you showed would be better? The only issue I see with that is there is no ping or notification that I've noticed when a page is changed.

moorepants commented:
You can submit the ideas here. Another way to approach:

1. Start a documentation page with the desired API with an example(s).
2. Write the tests based on agreed upon API in docs.
3. Implement the new classes/methods.

This is called "doc driven development" as opposed to "test driven". You can then use a PR and all material will end up in the source.

I could touch base briefly tomorrow. This is the last week of school here so I'm swamped with grading etc.

jbm950 commented:
Angadh and I plan on meeting Friday at 2PM (Pacific Time) if that would work better for you. I'll go ahead and work on the docs here unless you think it would be better to put them in a new PR. In planning an API for this it it is going to be necessary to include examples that use the formed equations of motion which means using other packages. I'm not sure that examples using other packages is going to be desired within the SymPy documentation.

moorepants commented:
What other packages are you talking about? In general, no we don't want any dependencies for SymPy, although numpy, scipy, theano, ipython, are optional dependencies for some modules.

jbm950 commented:
It wouldn't be dependencies for the suggested additions, just packages that use the output of the equations of motion generators. I haven't seen much in the SymPy repository that uses this output other than perhaps displaying it and so to develop an API for the output it would be useful to show how other packages would use it. Like a simulation that you presented in the "simple DAE problem" or some other example from PyDy.

moorepants commented:
Oh I see. Maybe it'd be better to make these full examples as a PR to pydy/examples then. Then you are free to use what you want.

moorepants commented:
This is being closed in favor of #11431.
